src/lib/operators/join_index.cpp

```diff
@@ -83,133 +83,130 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+
+      const auto& reference_segment_pos_list = reference_segment->pos_list();
+      if (reference_segment_pos_list->references_single_chunk()) {
+        if (reference_segment_pos_list->empty()) {
+          continue;
+        }
+        const auto& index_data_table_chunk = index_data_table->get_chunk((*reference_segment_pos_list)[0].chunk_id);
+        const auto& indexes = index_data_table_chunk->get_indexes(index_data_table_column_ids);
+        std::shared_ptr<BaseIndex> index = nullptr;
+
+        if (!indexes.empty()) {
+          // We assume the first index to be efficient for our join
+          // as we do not want to spend time on evaluating the best index inside of this join loop
+          index = indexes.front();
         }
-      }
-    }
-  }
 
-  // For Full Outer and Right Join we need to add all unmatched rows for the index side.
-  if (_mode == JoinMode::FullOuter || _mode == JoinMode::Right) {
-    for (ChunkID chunk_id{0}; chunk_id < _index_matches.size(); ++chunk_id) {
-      for (ChunkOffset chunk_offset{0}; chunk_offset < _index_matches[chunk_id].size(); ++chunk_offset) {
-        if (!_index_matches[chunk_id][chunk_offset]) {
-          _index_pos_list->emplace_back(RowID{chunk_id, chunk_offset});
-          _probe_pos_list->emplace_back(NULL_ROW_ID);
+        // Scan all chunks from the pruning side input
+        if (index) {
+          for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count();
+               ++probe_chunk_id) {
+            const auto& probe_segment = _probe_input_table->get_chunk(probe_chunk_id)
+                                            ->get_segment(_adjusted_primary_predicate.column_ids.first);
+            segment_with_iterators(*probe_segment, [&](auto probe_iter, const auto probe_end) {
+              _reference_join_two_segments_using_index(probe_iter, probe_end, probe_chunk_id, index_chunk_id, index,
+                                                       reference_segment_pos_list);
+            });
+          }
+          performance_data.chunks_scanned_with_index++;
+        } else {
+          _fallback_nested_loop(index_chunk_id, track_probe_matches, track_index_matches, is_semi_or_anti_join,
+                                secondary_predicate_evaluator);
         }
+      } else {
+        _fallback_nested_loop(index_chunk_id, track_probe_matches, track_index_matches, is_semi_or_anti_join,
+                              secondary_predicate_evaluator);
       }
     }
-  }
-
-  _probe_pos_list->shrink_to_fit();
-  _index_pos_list->shrink_to_fit();
-
-  // Write PosLists for Semi/Anti Joins, which so far haven't written any results to the PosLists
-  // We use `probe_matches_by_chunk` to determine whether a tuple from the probe side found a match.
-  if (is_semi_or_anti_join) {
-    const auto invert = _mode == JoinMode::AntiNullAsFalse || _mode == JoinMode::AntiNullAsTrue;
+  } else {  // INDEX DATA JOIN
```

Comment:
Is this branch also executed if the table type is reference but the join type is not inner? Is this a problem?

Reply:
That's the way it is supposed to be, so it's not a problem. The comment here may be misleading. This branch is for index data joins and for outer index reference joins.